sdk/core/Azure.Core/src/Pipeline/Internal/ReadTimeoutStream.cs

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Azure.Core.Pipeline
+{
+    internal class ReadTimeoutStream : ReadOnlyStream
+    {
+        private readonly Stream _stream;
+        private TimeSpan _readTimeout;
+        private CancellationTokenSource _cancellationTokenSource;
+
+        public ReadTimeoutStream(Stream stream, TimeSpan readTimeout)
+        {
+            _stream = stream;
+            _readTimeout = readTimeout;
+            UpdateReadTimeout();
+            _cancellationTokenSource = new CancellationTokenSource();
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            return _stream.Read(buffer, offset, count);
+        }
+
+        public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            var source = StartTimeout(cancellationToken, out bool dispose);
+            try
+            {
+                source.Token.Register(() =>
+                {
+                    Console.WriteLine("Cancelled");
+                });
+                return await _stream.ReadAsync(buffer, offset, count, source.Token).ConfigureAwait(false);
+            }
+            finally
+            {
+                StopTimeout(source, dispose);
+            }
+        }
+
+        private CancellationTokenSource StartTimeout(CancellationToken additionalToken, out bool dispose)
```

Comment:
This can be made a useful shared helper method, something like: 
```cs
public static IDisposable WithTimeout(this in CancellationToken token, TimeSpan timeout, out CancellationToken tokenWithTimeout)
{
    if (!token.CanBeCanceled && timeout.TotalMilliseconds < 0)
    {
        tokenWithTimeout = default;
        return Disposable.Empty;
    }

    CancellationTokenSource cts = CancellationTokenSource.CreateLinkedTokenSource(token);
    cts.CancelAfter(timeout);
    tokenWithTimeout = cts.Token;
    return cts;
}
```
And in `ReadAsync` it can be used as:
```cs
using IDisposable timeoutScope = cancellationToken.WithTimeout(_readTimeout, out var cancellationTokenWithTimeout);
try 
{
    return await _stream.ReadAsync(buffer, offset, count, cancellationTokenWithTimeout).ConfigureAwait(false);
}
catch (OperationCanceledException)
{
    Console.WriteLine("Cancelled");
    throw;
}
```

Reply:
True but it would result in way more allocations than the current option.

Reply:
If I understand it correctly, optimization with cached `CancellationTokenSource` will only work when `cancellationToken.CanBeCanceled` is `false`. Looks like a good case for object pool. Do we have one, btw?

Reply:
> will only work when cancellationToken.CanBeCanceled is false.

cancellationToken.CanBeCanceled == false is very common but it's not the only optimization. I don't have to allocate IDisposable per read too.

> Looks like a good case for object pool. Do we have one, btw?

I prefer to avoid the pool if I can avoid it and we don't have a general purpose poll atm.